File: library/utils/src/main/java/com/example/utils/network/UrlUtil.java
package com.example.utils.network;

import android.net.Uri;
import android.text.TextUtils;

import java.io.UnsupportedEncodingException;
import java.net.URLDecoder;
import java.net.URLEncoder;
import java.util.ArrayList;
import java.util.HashMap;
import java.util.List;
import java.util.Map;
import java.util.Set;

/**
 * URL操作类，例如：编解码操作
 */
public class UrlUtil {

    /**
     * 以UTF-8的编码形式 编码 http url
     */
    public static String encodeUrl(String url) {
        if (TextUtils.isEmpty(url)) {
            return "";
        }
        Uri uri = Uri.parse(url);
        if (uri == null) {
            return "";
        }
        // 拆分uri path，逐个编码
        List<String> pathSegments = uri.getPathSegments();
        List<String> encodedPathSegments = new ArrayList<>();
        for (String uriPathSegment : pathSegments) {
            if (TextUtils.isEmpty(uriPathSegment)) {
                continue;
            }
            try {
                encodedPathSegments.add(URLEncoder.encode(uriPathSegment, "UTF-8"));
            } catch (UnsupportedEncodingException e) {
                e.printStackTrace();
                return "";
            }
        }
        // 拆分uri parameters，逐个编码
        Map<String, String> encodedParameters = new HashMap<>();
        Set<String> parameterNames = uri.getQueryParameterNames();
        for (String parameterName : parameterNames) {
            if (TextUtils.isEmpty(parameterName)) {
                continue;
            }
            String parameterValue = uri.getQueryParameter(parameterName);
            if (TextUtils.isEmpty(parameterValue)) {
                parameterValue = "";
            }
            try {
                encodedParameters.put(parameterName, URLEncoder.encode(parameterValue, "UTF-8"));
            } catch (UnsupportedEncodingException e) {
                e.printStackTrace();
                return "";
            }
        }
        // 组合编码后的uri
        StringBuilder builder = new StringBuilder();
        builder.append(url.substring(0, url.indexOf(uri.getEncodedPath())));
        for (String uriEncodedPathSegment : encodedPathSegments) {
            if (TextUtils.isEmpty(uriEncodedPathSegment)) {
                continue;
            }
            builder.append("/");
            builder.append(uriEncodedPathSegment);
        }
        builder.append("?");
        for (Map.Entry<String, String> entry : encodedParameters.entrySet()) {
            if (entry == null) {
                continue;
            }
            builder.append(entry.getKey());
            builder.append("=");
            builder.append(entry.getValue());
            builder.append("&");
        }
        if (builder.lastIndexOf("&") == builder.length() - 1) {
            builder.delete(builder.lastIndexOf("&"), builder.length());
        }
        if (builder.lastIndexOf("?") == builder.length() - 1) {
            builder.delete(builder.lastIndexOf("?"), builder.length());
        }
        // 返回url
        return builder.toString();
    }

    /**
     * 以UTF-8的编码形式 解码 http url
     */
    public static String decodeUrl(String url) {
        if (TextUtils.isEmpty(url)) {
            return "";
        }
        Uri uri = Uri.parse(url);
        if (uri == null) {
            return "";
        }
        // 拆分uri path，逐个解码
        List<String> pathSegments = uri.getPathSegments();
        List<String> decodedPathSegments = new ArrayList<>();
        for (String uriPathSegment : pathSegments) {
            if (TextUtils.isEmpty(uriPathSegment)) {
                continue;
            }
            try {
                decodedPathSegments.add(URLDecoder.decode(uriPathSegment, "UTF-8"));
            } catch (UnsupportedEncodingException e) {
                e.printStackTrace();
                return "";
            }
        }
        // 拆分uri parameters，逐个解码
        Map<String, String> decodedParameters = new HashMap<>();
        Set<String> parameterNames = uri.getQueryParameterNames();
        for (String parameterName : parameterNames) {
            if (TextUtils.isEmpty(parameterName)) {
                continue;
            }
            String parameterValue = uri.getQueryParameter(parameterName);
            if (TextUtils.isEmpty(parameterValue)) {
                parameterValue = "";
            }
            try {
                decodedParameters.put(parameterName, URLDecoder.decode(parameterValue, "UTF-8"));
            } catch (UnsupportedEncodingException e) {
                e.printStackTrace();
                return "";
            }
        }
        // 组合解码后的uri
        StringBuilder builder = new StringBuilder();
        builder.append(url.substring(0, url.indexOf(uri.getEncodedPath())));
        for (String uriEncodedPathSegment : decodedPathSegments) {
            if (TextUtils.isEmpty(uriEncodedPathSegment)) {
                continue;
            }
            builder.append("/");
            builder.append(uriEncodedPathSegment);
        }
        builder.append("?");
        for (Map.Entry<String, String> entry : decodedParameters.entrySet()) {
            if (entry == null) {
                continue;
            }
            builder.append(entry.getKey());
            builder.append("=");
            builder.append(entry.getValue());
            builder.append("&");
        }
        if (builder.lastIndexOf("&") == builder.length() - 1) {
            builder.delete(builder.lastIndexOf("&"), builder.length());
        }
        if (builder.lastIndexOf("?") == builder.length() - 1) {
            builder.delete(builder.lastIndexOf("?"), builder.length());
        }
        // 返回url
        return builder.toString();
    }

}

File: library/utils/src/main/java/com/example/utils/network/UrlUtil.kt
package com.example.utils.network

import android.net.Uri
import android.text.TextUtils
import java.io.UnsupportedEncodingException
import java.net.URLDecoder
import java.net.URLEncoder
import java.util.*

/**
 * URL操作类，例如：编解码操作
 */
object UrlUtil {
    /**
     * 以UTF-8的编码形式 编码 http url
     */
    fun encodeUrl(url: String): String {
        val uri = Uri.parse(url) ?: return ""
        // 拆分uri path，逐个编码
        val pathSegments = uri.pathSegments
        val encodedPathSegments = ArrayList<String>()
        for (uriPathSegment in pathSegments) {
            try {
                encodedPathSegments.add(URLEncoder.encode(uriPathSegment, "UTF-8"))
            } catch (e: UnsupportedEncodingException) {
                e.printStackTrace()
                return ""
            }
        }
        // 拆分uri parameters，逐个编码
        val encodedParameters = HashMap<String, String>()
        val parameterNames = uri.queryParameterNames
        for (parameterName in parameterNames) {
            val parameterValue = uri.getQueryParameter(parameterName) ?: ""
            try {
                encodedParameters[parameterName] = URLEncoder.encode(parameterValue, "UTF-8")
            } catch (e: UnsupportedEncodingException) {
                e.printStackTrace()
                return ""
            }
        }
        // 组合编码后的uri
        val builder = StringBuilder()
        builder.append(url.substring(0, url.indexOf(uri.encodedPath!!)))
        for (uriEncodedPathSegment in encodedPathSegments) {
            builder.append("/")
            builder.append(uriEncodedPathSegment)
        }
        builder.append("?")
        for (entry in encodedParameters.entries) {
            builder.append(entry.key)
            builder.append("=")
            builder.append(entry.value)
            builder.append("&")
        }
        if (builder.lastIndexOf("&") == builder.length - 1) {
            builder.delete(builder.lastIndexOf("&"), builder.length)
        }
        if (builder.lastIndexOf("?") == builder.length - 1) {
            builder.delete(builder.lastIndexOf("?"), builder.length)
        }
        // 返回url
        return builder.toString()
    }

    /**
     * 以UTF-8的编码形式 解码 http url
     */
    fun decodeUrl(url: String): String {
        val uri = Uri.parse(url) ?: return ""
        // 拆分uri path，逐个解码
        val pathSegments = uri.pathSegments
        val decodedPathSegments: MutableList<String> =
            ArrayList()
        for (uriPathSegment in pathSegments) {
            try {
                decodedPathSegments.add(URLDecoder.decode(uriPathSegment, "UTF-8"))
            } catch (e: UnsupportedEncodingException) {
                e.printStackTrace()
                return ""
            }
        }
        // 拆分uri parameters，逐个解码
        val decodedParameters = HashMap<String, String>()
        val parameterNames = uri.queryParameterNames
        for (parameterName in parameterNames) {
            var parameterValue = uri.getQueryParameter(parameterName)
            if (TextUtils.isEmpty(parameterValue)) {
                parameterValue = ""
            }
            try {
                decodedParameters[parameterName] = URLDecoder.decode(parameterValue, "UTF-8")
            } catch (e: UnsupportedEncodingException) {
                e.printStackTrace()
                return ""
            }
        }
        // 组合解码后的uri
        val builder = StringBuilder()
        builder.append(url.substring(0, url.indexOf(uri.encodedPath!!)))
        for (uriEncodedPathSegment in decodedPathSegments) {
            builder.append("/")
            builder.append(uriEncodedPathSegment)
        }
        builder.append("?")
        for (entry in decodedParameters.entries) {
            builder.append(entry.key)
            builder.append("=")
            builder.append(entry.value)
            builder.append("&")
        }
        if (builder.lastIndexOf("&") == builder.length - 1) {
            builder.delete(builder.lastIndexOf("&"), builder.length)
        }
        if (builder.lastIndexOf("?") == builder.length - 1) {
            builder.delete(builder.lastIndexOf("?"), builder.length)
        }
        // 返回url
        return builder.toString()
    }
}